src/kronecker.jl

```diff
@@ -0,0 +1,258 @@
+Base.kron(A::LinearMap, B::LinearMap, Cs::LinearMap...) = KroneckerMap{promote_type(eltype(A), eltype(B), map(eltype, Cs)...)}(tuple(A, B, Cs...))
+Base.kron(A::AbstractMatrix, B::LinearMap) = kron(LinearMap(A), B)
+Base.kron(A::LinearMap, B::AbstractMatrix) = kron(A, LinearMap(B))
+# promote AbstractMatrix arguments to LinearMaps, then take LinearMap-Kronecker product
+for k in 3:8 # is 8 sufficient?
+    Is = ntuple(n->:($(Symbol(:A,n))::AbstractMatrix), Val(k-1))
+    # yields (:A1, :A2, :A3, ..., :A(k-1))
+    L = :($(Symbol(:A,k))::LinearMap)
+    # yields :Ak
+    mapargs = ntuple(n -> :(LinearMap($(Symbol(:A,n)))), Val(k-1))
+    # yields (:LinearMap(A1), :LinearMap(A2), ..., :LinearMap(A(k-1)))
+
+    @eval Base.kron($(Is...), $L, As::Union{LinearMap,AbstractMatrix}...) =
+        kron($(mapargs...), $(Symbol(:A,k)), promote_to_lmaps(As...)...)
+end
+
+promote_to_lmaps_(A::AbstractMatrix) = LinearMap(A)
+promote_to_lmaps_(A::LinearMap) = A
+promote_to_lmaps() = ()
+promote_to_lmaps(A) = (promote_to_lmaps_(A),)
+@inline promote_to_lmaps(A, B, Cs...) =
+    (promote_to_lmaps_(A), promote_to_lmaps_(B), promote_to_lmaps(Cs...)...)
+
+Base.size(A::KroneckerMap) = (prod(size.(A.maps, 1)), prod(size.(A.maps, 2)))
+
+LinearAlgebra.issymmetric(A::KroneckerMap) = all(issymmetric, A.maps)
+LinearAlgebra.ishermitian(A::KroneckerMap{<:Real}) = all(issymmetric, A.maps)
+LinearAlgebra.ishermitian(A::KroneckerMap) = all(ishermitian, A.maps)
+
+LinearAlgebra.adjoint(A::KroneckerMap{T}) where {T} = KroneckerMap{T}(map(adjoint, A.maps))
+LinearAlgebra.transpose(A::KroneckerMap{T}) where {T} = KroneckerMap{T}(map(transpose, A.maps))
+
+Base.:(==)(A::KroneckerMap, B::KroneckerMap) = (eltype(A) == eltype(B) && A.maps == B.maps)
+
+function A_mul_B!(y::AbstractVector, L::KroneckerMap{T,<:NTuple{2,LinearMap}}, x::AbstractVector) where {T}
+    # require_one_based_indexing(y)
+    (length(y) == size(L, 1) && length(x) == size(L, 2)) || throw(DimensionMismatch("A_mul_B!"))
+    A, B = L.maps
+    X = reshape(x, (size(B, 2), size(A, 2)))
+    _kronmul!(y, B, X, transpose(A), T)
+    return y
+end
+function A_mul_B!(y::AbstractVector, L::KroneckerMap{T}, x::AbstractVector) where {T}
+    # require_one_based_indexing(y)
+    (length(y) == size(L, 1) && length(x) == size(L, 2)) || throw(DimensionMismatch("A_mul_B!"))
+    A = first(L.maps)
+    B = kron(Base.tail(L.maps)...)
+    X = LinearMap(reshape(x, (size(B, 2), size(A, 2))); issymmetric=false, ishermitian=false, isposdef=false)
+    _kronmul!(y, B, X, transpose(A), T)
+    return y
+end
+# mixed-product rule, prefer the right if possible
+# (A₁ ⊗ A₂ ⊗ ... ⊗ Aᵣ) * (B₁ ⊗ B₂ ⊗ ... ⊗ Bᵣ) = (A₁B₁) ⊗ (A₂B₂) ⊗ ... ⊗ (AᵣBᵣ)
+function A_mul_B!(y::AbstractVector, L::CompositeMap{<:Any,<:Tuple{KroneckerMap,KroneckerMap}}, x::AbstractVector)
+    B, A = L.maps
+    if length(A.maps) == length(B.maps) && all(M -> check_dim_mul(M[1], M[2]), zip(A.maps, B.maps))
+        A_mul_B!(y, kron(map(prod, zip(A.maps, B.maps))...), x)
```

Comment:
You could probably also use `map(*, A.maps, B.maps)`, which is maybe somewhat more intuitive?